### PR TITLE
Sam/fix spend bug

### PR DESCRIPTION
**Pull request by samholmes**

### CODE REVIEW

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 1]
```diff
@@ -447,18 +430,37 @@ export function makeUtxoEngineState(
...
       for (const [scriptPubkey, utxos] of utxoMap.entries()) {
+        // Get saved utxo set
+        const savedUtxos = await processor.fetchUtxos({ scriptPubkey })
+        // Filter UTXOs to de-duplicate (and undefined utxo as a type assert)
+        const filteredUtxos = savedUtxos.filter(
+          utxo => utxo != null && !utxoIds.has(utxo?.id)
+        ) as IUTXO[]
+
+        // Add updated utxos to utxo set
+        const combinedUtxos = [...filteredUtxos, ...utxos]
```
Comment (thehobbit85) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:458]: Can a 'non-spent' utxo in appear in both the 'filteredUtxos' and the 'utxos' objects?
And if so, wouldn't that create duplications on 'combinedUtxos'?

**Review by thehobbit85 — CHANGES_REQUESTED** [review]
The first 3 commits are approved.
Regarding the last commit I would like to have a sidebar about it to better understand why and what you did here.

File: src/common/utxobased/engine/makeUtxoEngineState.ts (hunk 1 above)
Comment (samholmes) [reply src/common/utxobased/engine/makeUtxoEngineState.ts:458]: Yes, and I changed this last night to filter all IDs in the `utxos` set. This way there is no dupes.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 2]
```diff
@@ -447,18 +430,37 @@ export function makeUtxoEngineState(
...
+
+      // Map updated utxos
       for (const utxo of utxos) {
         const utxoSet = utxoMap.get(utxo.scriptPubkey) ?? []
         if (!utxoMap.has(utxo.scriptPubkey))
           utxoMap.set(utxo.scriptPubkey, utxoSet)
+        // Add utxo to set for scriptPubkey
         utxoSet.push(utxo)
+        // Add UTXO ID set
+        if (utxo.spent) utxoIds.add(utxo.id)
```
Comment (samholmes) [inline src/common/utxobased/engine/makeUtxoEngineState.ts]: Remove condition

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 3]
```diff
@@ -1221,25 +1223,69 @@ const processProcessorUtxos = async (
...
+        updatedUtxos[utxo.id] = { ...utxo, spent: true }
+      }
+    } else {
+      /*
+      If the UTXO is present in the updated UTXO set, enforce that the UTXO 
+      remains spent in the new UTXO set. This means, we will never mark a UTXO 
+      as unspent again. UTXOs only become unspent again if the tx is dropped
+      which shall be affected by another routine.
+      */
+      updatedUtxos[utxo.id].spent = utxo.spent || updatedUtxos[utxo.id].spent
```
Comment (samholmes) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:1265]: Add TODO comment

**Pull request merged**